### PR TITLE
ENH: Working on pretty-table summary for a list of headers

### LINKED ISSUE #61: add `summary` function
```
def summary(list_of_headers, *keys_to_show):
```

prints out a formatted prettytable with that info.

cc @tacaswell

### PR DESCRIPTION
**Pull request by ericdill**
Leaving this here so I don't forget about it.  Also so people can give
it a look and let me know right track or wrong track

Closes #61

### CODE REVIEW

File: databroker/broker.py [hunk 1]
```diff
@@ -217,10 +221,10 @@ def __call__(self, text_search=None, **kwargs):
 
         Examples
         --------
-        >>> DataBroker('keyword')  # full text search
-        >>> DataBroker(start_time='2015-03-05', stop_time='2015-03-10')
-        >>> DataBroker(data_key='motor1')
-        >>> DataBroker(data_key='motor1', start_time='2015-03-05')
+        >>> Broker('keyword')  # full text search
```
Comment (tacaswell) [inline databroker/broker.py]: Probably should not use `Broker` here is this needs to be an instance of this class, not the class it's self.

**Comment by tacaswell** [comment]
Seems like a reasonable direction.

Probably should lobby to have a mapping of descriptor uids to number of events added to the RunStop document.

**Comment by ericdill** [comment]
Demo of this working:
### Insert a specfile into metadatastore
```python
from suitcase import spec
specfilename = '20150323'
sf = spec.Specfile('/home/edill/ixs-data/{}'.format(specfilename))
ret = spec.insert_specfile_into_broker(sf)
spec.summarize_insertion(*ret)
```
### Retrieve all headers associated with that specfile and show a summary
```python
from databroker import broker, db

headers = list(db(specfilename=specfilename))
table = broker.summarize(headers, 'start-time', 'scan_id', 'uid-6', 'stop-uid-6', 'duration', 'num_events')

print(table)
```

Will print out:
```
+---------------------+---------+--------+------------+----------+----------------+
|      start-time     | scan_id | uid-6  | stop-uid-6 | duration |   num_events   |
+---------------------+---------+--------+------------+----------+----------------+
| 2015-03-30 14:52:00 |    13   | 4c593e |   73f060   | 0:00:00  | [1, 51, 1, 51] |
| 2015-03-30 10:52:10 |    12   | 7dff7f |   f3f8dd   | 0:00:00  | [1, 51, 1, 51] |
| 2015-03-27 18:09:48 |    9    | eb33a5 |   c6c257   | 0:00:00  | [1, 41, 1, 41] |
| 2015-03-27 09:33:31 |    8    | 8afb23 |   863390   | 0:00:00  | [1, 41, 1, 41] |
| 2015-03-27 09:21:15 |    7    | 2f5a3f |   b22b4d   | 0:00:00  | [1, 41, 1, 41] |
| 2015-03-26 16:29:41 |    6    | 9abb0e |   f76581   | 0:00:00  | [1, 41, 1, 41] |
| 2015-03-23 21:19:34 |    5    | d5e005 |   fab2d1   | 0:00:00  | [1, 51, 1, 51] |
| 2015-03-23 21:18:53 |    4    | cb397a |   63bfc4   | 0:00:00  |  [1, 2, 1, 2]  |
| 2015-03-23 20:26:43 |    3    | 002de8 |   f550a0   | 0:00:00  | [1, 41, 1, 41] |
| 2015-03-23 19:34:47 |    2    | 2477f4 |   b28577   | 0:00:00  | [1, 41, 1, 41] |
| 2015-03-23 18:39:18 |    1    | 94ad29 |   296fc1   | 0:00:00  | [51, 1, 1, 51] |
+---------------------+---------+--------+------------+----------+----------------+
```

Ah crap looks like my insertion code is not correctly identifying event descriptors. That's why there are two sets of baseline/primary events in the `num_events` column... :hankey:

**Comment by ericdill** [comment]
Oh great. Turns out that my last comment is easily fixed. I was using the hash() function again [elsewhere] (https://github.com/NSLS-II/suitcase/pull/17/files#diff-3073c31ff1ec8737f38e263e79571b9bR752) in the spec.py code

**Comment by ericdill** [comment]
There, now the table looks like this:

```
+---------------------+---------+--------+------------+----------+------------+
|      start-time     | scan_id | uid-6  | stop-uid-6 | duration | num_events |
+---------------------+---------+--------+------------+----------+------------+
| 2015-03-30 14:52:00 |    13   | f89482 |   ce7b2d   | 0:00:00  |  [1, 51]   |
| 2015-03-30 10:52:10 |    12   | 2a952a |   f2c7d4   | 0:00:00  |  [1, 51]   |
| 2015-03-27 18:09:48 |    9    | ff1c7e |   4e3bc3   | 0:00:00  |  [1, 41]   |
| 2015-03-27 09:33:31 |    8    | f5e63b |   b5c446   | 0:00:00  |  [1, 41]   |
| 2015-03-27 09:21:15 |    7    | b92291 |   bf7b3e   | 0:00:00  |  [1, 41]   |
| 2015-03-26 16:29:41 |    6    | 8e1f06 |   332cd8   | 0:00:00  |  [1, 41]   |
| 2015-03-23 21:19:34 |    5    | f891db |   2a4df9   | 0:00:00  |  [1, 51]   |
| 2015-03-23 21:18:53 |    4    | f58e20 |   35718e   | 0:00:00  |   [1, 2]   |
| 2015-03-23 20:26:43 |    3    | 2e0bbc |   3c5a31   | 0:00:00  |  [1, 41]   |
| 2015-03-23 19:34:47 |    2    | de6376 |   188ac3   | 0:00:00  |  [1, 41]   |
| 2015-03-23 18:39:18 |    1    | 3577ca |   648918   | 0:00:00  |  [51, 1]   |
+---------------------+---------+--------+------------+----------+------------+
```

when you run the following script:
```python

from suitcase import spec
specfilename = '20150323'
sf = spec.Specfile('/home/edill/ixs-data/{}'.format(specfilename))
ret = spec.insert_specfile_into_broker(sf)
spec.summarize_insertion(*ret)
from databroker import broker, db

headers = list(db(specfilename=specfilename))
table = broker.summarize(headers, 'start-time', 'scan_id', 'uid-6', 'stop-uid-6', 'duration', 'num_events')

print(table)
```
(assuming you have a specfile on your local computer, just change the specfilename to point to it, and obviously change the path...)

**Comment by ericdill** [comment]
Ok this is ready for review @tacaswell @danielballan @klauer

File: databroker/broker.py [hunk 2]
```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
...
+def _(header, key, default_value):
+    document = header
+    if 'uid' in key:
+        return _get_uid(key, document)
+    if key == 'time':
+        return datetime.fromtimestamp(document.get(key, default_value))
+    return document.get(key, default_value)
+
+
+def summarize(headers, *keys, default_value="N/A"):
```
Comment (tacaswell) [inline databroker/broker.py]: This is not valid LPy.  Need to use kwrags, pop the one key you want and then check that it is empty.

File: databroker/broker.py (hunk 1 above)
Comment (tacaswell) [inline databroker/broker.py]: bump on this.

File: databroker/broker.py [hunk 3]
```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
...
+        # see if we have a hyphen
+        key, num = key.split('-')
+        num = int(num)
+    except ValueError:
+        # Nope! Must be asking for the whole uid
+        num = None
+    return document.uid[:num]
+
+
+@singledispatch
```
Comment (tacaswell) [inline databroker/broker.py]: I do not see what the single dispatch is helping with here.  In `summarize` it looks like you should only ever see `Headers` and in the `Header` `_get_value` it only every sees `Documents`.

**Comment by tacaswell** [comment]
Can you rebase this so that gh hides the diffs from the pytest refactor?

File: databroker/broker.py (hunk 2 above)
Comment (ericdill) [inline databroker/broker.py]: In that case I'll just make keys a required argument that is an iterable

File: databroker/broker.py (hunk 3 above)
Comment (ericdill) [inline databroker/broker.py]: You're right. I'll refactor to remove singledispatch

File: databroker/broker.py [hunk 4]
```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
...
+def _get_value(h, key, default_value):
+    """
+    Get the value specified by `key` from the Header `h`. Give the value
+    of `default_value` for keys that are not in `h`
+
+    Parameters
+    ----------
+    h : databroker.core.Header
+        A single header to try and get a key from
+    key : str
```
Comment (licode) [inline databroker/broker.py]: Is this key similar to data_keys in descriptor?

File: databroker/broker.py [hunk 5]
```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
...
+        The key to find a value for in `h`. See the `*keys` section of the
+        `databroker.broker.summarize` docstring
+    default_value : str
+        The default value for which to fill missing keys
```
Comment (licode) [inline databroker/broker.py]: More explanation will be helpful. Why do you need to set a value to a key if it is not included in hdr originally?

File: databroker/broker.py [hunk 6]
```diff
@@ -562,3 +566,130 @@ def _munge_time(t, timezone):
...
+        Keys that exist in the RunStart, Descriptor or Stop documents. Or one
+        of the entries in the `known_special_keys` list.
+
+        Specify which document to get a key from by providing the document
+        name, followed by a hyphen, followed by the key. If no document name
+        is provided, it is assumed you are asking for the key in the run
+        start document. e.g.:
+            start-<key>: get <key> from header.start
+            stop-<key>: get <key> from header.stop
+            descriptor-<key>: get <key> from the descriptors
```
Comment (licode) [inline databroker/broker.py]: Here I get better ideas on how to use keys. 

👍 in seeing the implementation of this function, should be very useful!

**Comment by danielballan** [comment]
Nice feature. I thought hard about bike-shedding the contents of that table, but I have to agree that that is exactly the right set of things to show.

File: databroker/broker.py (hunk 5 above)
Comment (ericdill) [inline databroker/broker.py]: If you have multiple headers, some that have keys and some that don't, you will want a default value

**Comment by tacaswell** [comment]
Moved to #72

**Pull request closed without merging**